### PR TITLE
Struct and enum declarations no longer go through finalizeDecl()

### PR DESCRIPTION
**Pull request by slavapestov**
Builds on https://github.com/apple/swift/pull/26119. Now that getStoredProperties() is a request, Sema no longer has to finalize struct and enum declarations.

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please test compiler performance

**Comment by slavapestov** [comment]
(Both source compat failures are UPASS because an unrelated problem was fixed on master)

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

File: lib/SIL/TypeLowering.cpp [hunk 1]
```diff
@@ -237,7 +237,7 @@ namespace {
       llvm_unreachable("shouldn't get an inout type here");
     }
     RetTy visitErrorType(CanErrorType type) {
-      llvm_unreachable("shouldn't get an error type here");
```
Comment (CodaFi) [inline lib/SIL/TypeLowering.cpp]: How are error types wandering into SIL type lowering?

Comment (slavapestov) [reply lib/SIL/TypeLowering.cpp]: Type checking is now sufficiently lazy that if a struct has a field with an invalid type, and you don't directly reference the invalid field from source, SILGen might end up lazily validating the stored properties of the struct anyway, if you nevertheless emit an expression involving a value of the struct's type.

**Comment by DougGregor** [comment]
This is fantastic!

**Comment by DougGregor** [comment]
Do we get detection of accidentally-recursive struct/enum definitions for free now from type layout?

**Pull request merged**